core/src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -682,14 +745,100 @@ decodeUtf8Lax = decodeUtf8
+
+    handleError err souldBackTrack =
+        case cfm of
+            ErrorOnCodingFailure -> ParserD.Error err
+            TransliterateCodingFailure ->
+                case souldBackTrack of
+                    True -> ParserD.Done 1 replacementChar
+                    False -> ParserD.Done 0 replacementChar
+            DropOnCodingFailure ->
+                case souldBackTrack of
+                    True -> ParserD.Continue 1 UTF16CharDecodeInit
+                    False -> ParserD.Continue 0 UTF16CharDecodeInit
+
+    {-# INLINE combineSurrogates #-}
+    combineSurrogates hi lo =
+        let first10 = fromIntegral (hi - utf16HighSurrogate) `shiftL` 10
+            second10 = fromIntegral (lo - utf16LowSurrogate)
+         in unsafeChr (0x10000 + (first10 .|. second10))
+
+    -- In little endian we should get low surrogate first.
+    {-# INLINE step #-}
+    step UTF16CharDecodeInit x
+         -- XXX Combine the cases and check perf
+         | x < 0xD800 = pure $ ParserD.Done 0 $ unsafeChr $ fromIntegral x
+         | x > 0xDFFF = pure $ ParserD.Done 0 $ unsafeChr $ fromIntegral x
+         | x >= 0xD800 && x <= 0xDBFF = -- high surrogate
+               pure $ ParserD.Partial 0 $ UTF16CharDecoding x
+         | otherwise =
+               let msg = prefix ++ "Invalid first UTF16 word " ++ show x
+                in pure $ handleError msg False
+
+    -- We backtrack here as there is a chance the word in the buffer will
+    -- combine with the word after the current word.
+    step (UTF16CharDecoding prev) x
+         | x >= 0xDC00 && x <= 0xDFFF = -- low surrogate
+               pure $ ParserD.Done 0 $ combineSurrogates prev x
+         | otherwise =
+               let msg = prefix
+                       ++ "Invalid subsequent UTF16 word " ++ show x
+                       ++ " in state " ++ show prev
+                in pure $ handleError msg True
+
+    {-# INLINE extract #-}
+    extract UTF16CharDecodeInit =  error $ prefix ++ "Not enough input"
+    extract (UTF16CharDecoding _) =
+        case cfm of
+            ErrorOnCodingFailure ->
+                return $ ParserD.Error $ prefix ++ "Not enough input"
+            TransliterateCodingFailure ->
+                return (ParserD.Done 0 replacementChar)
+            -- XXX We shouldn't error out here. There is no way to represent an
+            -- empty parser result unless we return a "Maybe" type.
+            DropOnCodingFailure -> error $ prefix ++ "Not enough input"
+
 -- | Decode a UTF-16 little endian encoded bytestream to a stream of Unicode
 -- characters. The function throws an error if an invalid codepoint is
 -- encountered.
 --
--- /Unimplemented/
-{-# INLINE decodeUtf16le' #-}
-decodeUtf16le' :: Stream m Word16 -> Stream m Char
-decodeUtf16le' = undefined
+{-# INLINE decodeUtf16' #-}
+decodeUtf16' :: Monad m => Stream m Word16 -> Stream m Char
+decodeUtf16' =
+    fmap fromEither
+        . Stream.parseMany (parseCharUtf16With ErrorOnCodingFailure)
```

Comment:
How do you handle empty stream? Do we have a test for that? The parser would error out. You can try alternative parser but need to check perf.

Reply:
Hmm. We cannot use a parser then. This should be some sort of a stateful transformation.